### DISCUSSION TITLE
사소하지만 재밌는 CS! 패턴 찍어먹기

### DISCUSSION THREAD

**Discussion by byhhh2** [woowacourse-study/2022-woowahan-bansanghwe#147] [JavaScript]
# 디자인 패턴과 프로그래밍 패러다임

디자인 패턴이란 프로그램을 설계할 때 발생했던 **문제점**들을 객체간의 상호 관계 등을 이용하여 해결할 수 있도록 **하나의 규약 형태로 만들어 놓은 것**을 의미한다. 

## 싱글톤 패턴

하나의 클래스에 오직 하나의 인스턴스만 가지는 패턴 

- 데이터베이스 연결 모듈에 많이 사용한다.
- 하나의 인스턴스를 만들어 놓고 해당 인스턴스를 다른 모듈들이 공유하며 사용한다.
- 자바스크립트에서 리터럴(`{}`) 혹은 `new Object`로 객체를 생성하면 다른 어떤 객체와도 같지 않기 때문에 이 자체만으로 싱글톤 패턴 구현 가능

```jsx
// 메인 모듈 
const mysql = require('mysql');
const pool = mysql.createPool({});
pool.connect();

// 모듈 A
pool.query(query, () => {});

// 모듈 B
pool.query(query, () => {});

// 메인 모듈에서 데이터베이스 연결에 관한 인스턴스를 정의하고
// 다른 모듈인 A 또는 B에서 해당 인스턴스를 기반으로 쿼리를 보내는 형식 
```

### 장점

- 인스턴스 생성 비용 감소
    - 데이터베이스 연결에 대한 인스턴스 생성 비용(connection을 여는 것 등)을 아낄 수 있다.

### 단점

- 의존성 증가
- TDD에 걸림돌이 된다.
    - TDD는 단위 테스트를 주로 한다. 단위 테스트는 테스트가 독립적이어야 하며 테스트를 어떤 순서로든 실행할 수 있어야 한다. 하지만 싱글톤은 하나의 인스턴스를 기반으로 하므로 테스트마다 독립적인 인스턴스를 만들기 어렵다.

> **의존성**
> 
> 
> 
> A가 B에 의존성이 있다. = B의 변경 사항에 A 또한 변해야 된다. 
> 
> **의존성 주입**
> 
> 싱글톤 패턴의 단점은 모듈 간의 결합을 강하게 만들 수 있다는 점이다. 
> 
> 이때 의존성 주입 (DI, Dependency Injection)을 통해 모듈간의 결합을 좀 더 느슨하게 만들 수 있다. 
> 
> <img width="366" alt="스크린샷 2022-12-26 오후 6 55 19" src="https://user-images.githubusercontent.com/52737532/209618510-0054fae2-394d-4003-9841-4091b60ef5cb.png">
>  
> **메인 모듈**이 직접 다른 하위 모듈에 대한 의존성을 주기 보다는
> 
> **의존성 주입자가** 이부분을 가로채 메인 모듈이 간접적으로 의존성을 주입하는 방식 
> 
> 의존성 주입의 **장점**
> 
> - 모듈을 쉽게 교체할 수 있는 구조가 되어 테스팅이 쉽고, 마이그레이션이 수월하다.
> - 의존성 방향이 일관된다. (추상화 레이어를 넣고 ← 구현체를 넣음)
> - 모듈 간 관계가 명확해진다.
> 
> 의존성 주입의 **단점** 
> 
> - 모듈 분리가 많아져 복잡성 증가
> 
> 의존성 주입의 **원칙** 
> 
> - 상위 모듈은 하위 모듈에서 어떤 것도 가져오지 말아야 한다.
> - 추상화에 의존해야 한다.
> - 추상화는 세부사항에 의존하지 말아야 한다.

## 팩토리 패턴

객체 생성 부분을 떼어내 추상화한 패턴

- 상속 관계가 있는 두 클래스에서 **상위 클래스가 중요한 뼈대를 결정**하고 
**하위 클래스에서 객체 생성에 관한 구체적인 내용을 결정**한다.
- 상위 클래스와 하위 클래스가 분리되어 느슨한 결합을 가진다. (유연성 증가)
- 객체 생성 로직이 떨어져 있어 유지 보수성 증가
- **라떼, 아메리카노, 우유 레시피 (구체적인 내용)이 들어있는 하위 클래스**가 컨베이어 벨트를 통해 전달되고, **상위 클래스인 바리스타 공장**에서 이 레시피들을 토대로 생산하는 공정
- 의존성 주입으로도 볼 수 있다.

<img width="399" alt="스크린샷 2022-12-26 오후 7 28 29" src="https://user-images.githubusercontent.com/52737532/209618538-7e33720c-e3d8-4429-9951-db611241fac5.png">

```jsx
// 전달받은 값에 따라 다른 객체를 생성하며 인스턴스 타입 등을 정한다. 

const num = new Object(42);
const str = new Object('abc');

num.constructor.name; // Number
str.constructor.name; // String
```

```jsx
class Latte { }
class Espresso { }

class LatteFactory { // 하위 클래스 
	static createCoffee() {
		return new Latte(); 
	}
} 

class EspressoFactory { // 하위 클래스 
	static createCoffee() {
		return new Espresso(); 
	}
}

const factoryList = { LatteFactory, EspressoFactory };

class CoffeeFactory { // 상위 클래스 
	static createCoffee(type) {
		const factory = factoryList[type];
		return factory.createCoffee();
	}
}

// main

const coffee = CoffeeFactory.createCoffee('LatteFactory');
```

## 전략 패턴

객체의 행위를 바꾸고 싶은 경우 직접 수정하지 않고 **전략이라고 부르는 캡슐화한 알고리즘**을 컨텍스트 안에서 바꿔주며 상호 교체가 가능하게 만드는 패턴 

- 컨텍스트 : 개발자가 어떤 작업을 완료하는 데 필요한 모든 관련 정보

<img width="222" alt="스크린샷 2022-12-26 오후 8 03 17" src="https://user-images.githubusercontent.com/52737532/209618574-1fbbf5d2-39d4-44b5-af3e-8a55b077ad62.png">

- 전략만 변경해서 인증하는 경우 (예시 passport 라이브러리)
    - 서비스 내 회원가입된 아이디/비밀번호 기반으로 인증하는 LocalStrategy와 페이스북, 네이버 등 다른 서비스를 기반으로 인증하는 OAuth 전략

```jsx
// 상품을 구매할 때 네이버페이, 카카오페이 등 다양한 방법으로 결제할 수 있다.
// 결제 방식 전략 변경 

cart.pay(new NaverCard());
cart.pay(new KakaoCard());
```

```jsx
let passport = require('passport');

passport.use(new LocalStrategy());

// 전략을 use의 매개변수로 넣어서 로직을 수행 
```

## 옵저버 패턴

주체가 어떤 객체의 상태 변화를 관찰하다가 상태 변화가 있을 때 메서드 등을 통해 옵저버 목록에 있는 옵저버들에게 변화를 알려주는 디자인 패턴 

- 주체 : 관찰자
- 옵저버들 : 객체의 상태 변화에 따라 추가 변화 사항이 생기는 객체들
- 옵저버 패턴을 활용한 서비스로는 트위터가 있다.
    - 새로운 트윗이 있을 경우 팔로우한 옵저버들에게 이를 알려준다.
- 주로 이벤트 기반 시스템에 사용된다. MVC 패턴에 사용되기도 한다.
    - MVC : model에 변경사항이 생겨 이를 옵저버인 view에 알려주고 이를 기반으로 controller가 작동한다.

<img width="302" alt="스크린샷 2022-12-26 오후 8 10 27" src="https://user-images.githubusercontent.com/52737532/209618607-fe0ec694-e2eb-408d-b6ba-3189b0ab2ebd.png">

<img width="464" alt="스크린샷 2022-12-26 오후 8 12 16" src="https://user-images.githubusercontent.com/52737532/209618622-9ebf7c58-9ba1-47ab-b8dc-83228b4a9aae.png">

> **프록시 객체**
> 
> 
> 
> 어떤 대상의 기본적인 특성의 작업을 가로챌 수 있는 객체
> 
> 자바스크립트에서 프록시 객체는 두가지 매개변수를 가진다. 
> 
> - target : 프록시할 대상
> - handler : 프록시 객체의 target 동작을 가로채서 정의할 동작들이 정해져 있는 함수
> 
> ```jsx
> const handler = {
> 	get: function(target, name) {
> 		return name === 'name' ? `${target.a} ${target.b}` : target[name];
> 	}
> }
> 
> const p = new Proxy({ a: 'hello', b: 'world' }, handler);
> 
> console.log(p.name); // hello world 
> 
> // name이라는 속성에 접근할 때는 a와 b를 합쳐서 문자열을 만들어라. 
> 
> // p라는 변수에 name이라는 속성을 선언하지 않았는데도, 
> // p.name으로 name 속성에 접근하려고 할 때 그 부분을 가로챈다. 
> ```
> 

### 프록시 객체를 이용한 옵저버 패턴

```jsx
function createReactiveObject(target, callback) {
	const proxy = new Proxy(target, {
		set: function(obj, prop, value) {
					if (value !== obj[prop]) {
						const prevValue = obj[prop];
						obj[prop] = value;
						callback(`${prop}가 [${prevValue}] >> [${value}]로 변경되었습니다.`);
					}
		
					return true;
				}
	})

	return proxy;
}

const obj = {
	'인사': '하이',
};

const p = createReactiveObject(obj, console.log);

p['인사'] = '바이';

// set 함수를 통해 속성에 대한 접근을 가로채 '인사' 속성이 
// 하이 -> 바이로 되는 것을 감시할 수 있다. 
```

- Vue에서는 ref로 정의하면 해당값이 변경되었을 때 자동으로 DOM에 있는 값이 변경되는 데, 이는 프록시 객체를 이용한 옵저버 패턴을 이용하여 구현한 것

### 프록시 패턴

대상 객체에 접근하기 전 그 접근에 대한 흐름을 가로채 대상 객체 앞단의 인터페이스 역할을 하는 디자인 패턴 

<img width="416" alt="스크린샷 2022-12-26 오후 8 50 23" src="https://user-images.githubusercontent.com/52737532/209618636-1827096c-f884-43c2-b3f7-55c5d37c26bb.png">

- 객체 변경을 보완하며 보안, 데이터 검증, 캐싱, 로깅에 사용한다.

### 프록시 서버

서버와 클라이언트 사이에서 클라이언트가 자신을 통해 다른 네트워크 서비스에 간접적으로 접속할 수 있게 해주는 컴퓨터 시스템이나 응용 프로그램 

- 예시는 nginx
    - 주로 nodejs 서버 앞단의 프록시 서버로 활용된다.
    - 익명 사용자의 직접적인 서버로의 접근을 차단하고 간접적으로 한 단계를 더 거침으로써 보안성을 강화한다.
    - 실제 포트를 숨기고, 정적 자원을 gzip 압축하거나, 메인 서버 앞단에서의 로깅을 할 수도 있다.

<img width="499" alt="스크린샷 2022-12-26 오후 8 55 20" src="https://user-images.githubusercontent.com/52737532/209618664-336db736-d57e-4473-95c2-1f4ead84bfc5.png">

### CORS와 프론트앤드 프록시 서버

- CORS : 서버가 웹 브라우저에서 리소스를 로드할 때 다른 오리진을 통해 로드하지 못하게 하는 HTTP 헤더 기반 메커니즘

**Comment by airman5573** [comment]
꼼꼼한 정리 믓찌다 👍
